sw/src/service-worker/service-worker.js

```diff
@@ -0,0 +1,420 @@
+        if (!event.data) return
+
+        // This allows the web app to trigger skipWaiting via
+        // registration.waiting.postMessage({type: 'SKIP_WAITING'})
+        // Paired with `clientsClaim()` at top of file.
+        if (event.data.type === 'SKIP_WAITING') {
+            self.skipWaiting()
+        }
+
+        if (event.data.type === 'START_RECORDING') {
+            startRecording(event)
+        }
+
+        if (event.data.type === 'COMPLETE_RECORDING') {
+            completeRecording(event.source.id) // same as FetchEvent.clientId
+        }
+    })
+
+    // Open DB on activation
+    self.addEventListener('activate', event => {
+        event.waitUntil(createDB().then(removeUnusedCaches))
+    })
+
+    // * 3. Helper functions:
+
+    function urlMeetsDefaultCachingCriteria(url) {
+        // Don't cache if pwa.caching.omitExternalRequests in d2.config is true
+        if (OMIT_EXTERNAL_REQUESTS && url.origin !== self.location.origin)
+            return false
+
+        // Don't cache if url matches filter in pattern list from d2.config.js
+        const urlMatchesFilter = URL_FILTER_PATTERNS.some(pattern =>
+            new RegExp(pattern).test(url.pathname)
+        )
+        if (urlMatchesFilter) return false
+
+        return true
+    }
+
+    /** Called upon SW activation */
+    function createDB() {
+        dbPromise = openSectionsDB()
+        return dbPromise
+    }
+
+    /** Called upon SW activation */
+    async function removeUnusedCaches() {
+        const cacheKeys = await caches.keys()
+        return Promise.all(
+            cacheKeys.map(async key => {
+                const isWorkboxKey = /workbox/.test(key)
+                const isInKeepList = !!CACHE_KEEP_LIST.find(
+                    keepKey => keepKey === key
+                )
+                const db = await dbPromise
+                const isASavedSection = !!(await db.get(SECTIONS_STORE, key))
+                if (!isWorkboxKey && !isInKeepList && !isASavedSection) {
+                    console.debug(
+                        `[SW] Cache with key ${key} is unused and will be deleted`
+                    )
+                    return caches.delete(key)
+                }
+            })
+        )
+    }
+
+    // Triggered on 'START_RECORDING' message
+    function startRecording(event) {
+        console.debug('[SW] Starting recording')
+        if (!event.data.payload?.sectionId)
+            throw new Error('[SW] No section ID specified to record')
+
+        const clientId = event.source.id // clientId from MessageEvent
+        // Throw error if another recording is in process
+        if (isClientRecording(clientId))
+            throw new Error(
+                "[SW] Can't start a new recording; a recording is already in process"
+            )
+
+        const newClientRecordingState = {
+            sectionId: event.data.payload?.sectionId,
+            pendingRequests: new Map(),
+            fulfilledRequests: new Map(),
```

Comment:
These are `Map`s because I expected to add response data to the URLs and add them to the IndexedDB, but response objects [can't get saved in IndexedDB](https://stackoverflow.com/questions/32880073/whats-the-best-option-for-structured-cloning-of-a-fetch-api-request-object). These two 'requests' objects could be `Set`s for now, but `Map` might be useful when we're normalizing response data